### ISSUE TITLE
ocamltest/summarize.awk not properly reporting abort failures on testsuite runs

### ISSUE THREAD

**Issue by abbysmal** [ocaml/ocaml#10953]
@kayceesrk mentioned to the Multicore team earlier today that something strange could be seen running the testsuite on trunk, with specific parameters.

After investigating a bit, I found out that `summarize.awk` and `ocamltest` can skip reporting abort failures because the log can be improperly formatted, see this example:

`OCAMLRUNPARAM="s=4096,v=0" USE_RUNTIME=d make one TEST=tests/tmc//'partial_application.ml' > partial_application.txt`

`awk -f summarize.awk partial_application.txt`

Will announce the test as passed.

But we can clearly see a failure in the log in question
 ```
... testing 'partial_application.ml' with 2 (native) => Process 746304 got signal 6(Aborted), core dumped                                                                
Could not find core file.                                                                                                                                                 
failed (Running program /home/engil/Documents/Dev/ocaml/testsuite/_ocamltest/tests/tmc/partial_application/ocamlopt.byte/partial_application.opt without any argument: command
```

This seems to be related to the fact that instead of formatting the line properly and having the correct `(native) => failed`, we have some logs resulting from the process aborting. (and we can see the correct `failed` line being inserted a bit later in the log.)

I am not sure how to go about fixing this specific issue, I'm not well versed in `awk`, and how `ocamltest` can be made to avoid logging such lines.

Also, cc @dra27 @shindere who may know how to address this.

**Comment by dra27** [comment]
Just in case we fix the underlying bug before the testsuite is addressed, ef0e05f645f6413c7299b8b38f40d3242091cf34 definitely exhibits this with the command given above.

**Comment by gasche** [comment]
An obvious, naive suggestion: we could stop driving the makefiles with Makefile+awk, and teach `ocamltest` to run test directories, not just tests, and print a report in a robust way. Are there good reasons *not* to do this? Are there alternatives that work with less effort?

Potential reasons not to do this:
- Maybe we are worried about the robustness of an ocamltest driver / the fact that it depends on OCaml itself. But it seems to work fine so far, and we trust it for each test individually already.
- Maybe @shindere is working on something else / not willing to go in this direction right now. It sounds like something that someone else could contribute if there was desire/interest.
- Someone wrote a `make parallel` target and then complained to have it fixed for Multicore, how would ocamltest do that? I actually suspect that threads would work fine. (We probably don't need a multi-domain program as most of the driver time would be spent waiting for commands to complete.)

Potential alternatives to this approach:
- I guess that ocamltest already produces test output in some per-test structured format (stored inside `_ocamltest/...`); we could write a script (in OCaml or awk or whatever) that processes this data, instead of parsing the output. (Question: how to distinguish data produced by separate test runs? We could store the data of each run in a new/free output directory?)

**Comment by dra27** [comment]
I don't think there are any good reasons not to do what you suggest beyond the time required to do it 🙂 I expect I will more quickly fix the awk script (or possibly ocamltest to put the `=> failed` first) than a full overhaul of the logging system, but I'd really like the latter... apart from anything else, it would be good to be able to summarise, say, all the test results in precheck as one, and I have no intention of ever doing that in awk!!

**Comment by xavierleroy** [comment]
> we could stop driving the makefiles with Makefile+awk, and teach ocamltest to run test directories, not just tests, and print a report in a robust way. Are there good reasons not to do this?

One of the self-inflicted constraints for implementing ocamltest was to not rely on the Unix library, or as little as possible, since the Unix library is to be considered buggy until it is tested using ocamltest :-)  Same thing for Threads, Str, etc.

This means that reimplementing GNU parallel inside ocamltest is not possible, as you'd need quite a bit of the Unix library.  (You could do without Threads, though.)  Plus, GNU parallel is quite clever and reimplementing it feels like a waste of time.

Running directories from within ocamltest might be possible, using stdlib functions only.

Parsing the logs and producing the report could possibly be done by a separate OCaml program that would be more robust than the current awk scripts.  As @dra27 wrote it's a matter of how much time you want to spend on that.

**Comment by gasche** [comment]
I think that a more robust design than what we currently have should avoid parsing text outputs produced by a mixture of the testing tool and the underlying texts.

If we don't want ocamltest to drive the tests, we could:
- Have a test-result-aggregate format that is stored as a serialized OCaml data structure.
- Have each ocamltest run produce this format in a "result file" for each test directory (or test file), in a standard place the Makefile driver knows about.
- Have the Makefile driver call ocamltest with all result files for summarization: `ocamltest -summarize -args0 list-of-result-files`.

This design keeps the current driver logic, in particular we can use `parallel` in the same way as before, with no parsing of results.

**Comment by xavierleroy** [comment]
> Have a test-result-aggregate format that is stored as a serialized OCaml data structure.

The serializer is something we need to test thoroughly using ocamltest...  Will you please agree to use the absolute strict minimum of code from the OCaml runtime system and standard library to implement ocamltest ?  And also that we have more urgent things to do ?

**Comment by shindere** [comment]
Enguerrand (2022/01/26 11:13 -0800):
> @kayceesrk mentioned to the Multicore team earlier today that something strange could be seen running the testsuite on trunk, with specific parameters.
>
> After investigating a bit, I found out that `summarize.awk` and `ocamltest` can skip reporting abort failures because the log can be improperly formatted, see this example:
>
> `OCAMLRUNPARAM="s=4096,v=0" USE_RUNTIME=d make one TEST=tests/tmc//'partial_application.ml' > partial_application.txt`
>
> `awk -f summarize.awk partial_application.txt`
>
> Will announce the test as passed.
>
> But we can clearly see a failure in the log in question
>  ```
> ... testing 'partial_application.ml' with 2 (native) => Process 746304 got signal 6(Aborted), core dumped
> Could not find core file.
> failed (Running program /home/engil/Documents/Dev/ocaml/testsuite/_ocamltest/tests/tmc/partial_application/ocamlopt.byte/partial_application.opt without any argument: command
> ```
>
> This seems to be related to the fact that instead of formatting the
> line properly and having the correct `(native) => failed`, we have
> some logs resulting from the process aborting. (and we can see the
> correct `failed` line being inserted a bit later in the log.)

Yes. That printing is done by C code in ocamltest, see the
`handle_process_termination` funciton in `ocamltest/run_unix.c`.

> I am not sure how to go about fixing this specific issue, I'm not well
> versed in `awk`, and how `ocamltest` can be made to avoid logging such
> lines.

I'm more concerned with the second problem and would be happy to hear
any suggestion. Should the stub code write to a file that ocamltest can
then read andinclude inits own log? Or should ocamltest flush and close
its log file before calling C code so that the C code can happen to the
log? Or should the Ccode print to a buffer in memory?

Really I don't knowwhat would be the best way to go but if we can agree
on a solution I'd be more than happy to submit a PR to fix ocamltest,
leaving theawk bit to somebody more motivated than me.

**Comment by shindere** [comment]
David Allsopp (2022/02/02 15:51 +0000):
> I don't think there are any good reasons not to do what you suggest
> beyond the time required to do it 🙂 I expect I will more quickly fix
> the awk script (or possibly ocamltest to put the `=> failed` first)
> than a full overhaul of the logging system, but I'd really like the
> latter... apart from anything else, it would be good to be able to
> summarise, say, all the test results in precheck as one, and I have no
> intention of ever doing that in awk!!

Given that the superfluous output is produced by C code, I'm unsure how
to do to printthe `=> failed` first but willing to implement it if
somebody suggests a solution. I'm also happy to review any PR on that if
I am Cc'ed.

**Comment by shindere** [comment]
Xavier Leroy (2022/02/02 08:16 -0800):
> > we could stop driving the makefiles with Makefile+awk, and teach ocamltest to run test directories, not just tests, and print a report in a robust way. Are there good reasons not to do this?
>
> One of the self-inflicted constraints for implementing ocamltest was to not rely on the Unix library, or as little as possible, since the Unix library is to be considered buggy until it is tested using ocamltest :-)  Same thing for Threads, Str, etc.
>
> This means that reimplementing GNU parallel inside ocamltest is not
> possible, as you'd need quite a bit of the Unix library.  (You could
> do without Threads, though.)  Plus, GNU parallel is quite clever and
> reimplementing it feels like a waste of time.

I'm not sure why we need GNU parallel rather than make?

The way I imagined things, we oculd somehwo collect all the tests to run
in a Make varialbe ahd have a target depending on allof them and then
`make -j` would do the jub (literally).

What would be wrong with that?

> Running directories from within ocamltest might be possible, using
> stdlib functions only.

Indeed.

> Parsing the logs and producing the report could possibly be done by a
> separate OCaml program that would be more robust than the current awk
> scripts.  As @dra27 wrote it's a matter of how much time you want to
> spend on that.

In my opinion, it's rather amatter of producing the output of ocamltest
in a format for whichthere are already parsers. Stillneed to check what
is available that could work intheocntext of GitHub actions andon a
Jenkins-based CI.

**Comment by shindere** [comment]
Gabriel Scherer (2022/02/02 08:25 -0800):
> I think that a more robust design than what we currently have should avoid parsing text outputs produced by a mixture of the testing tool and the underlying texts.
>
> If we don't want ocamltest to drive the tests, we could:
> - Have a test-result-aggregate format that is stored as a serialized OCaml data structure.
> - Have each ocamltest run produce this format in a "result file" for each test directory (or test file), in a standard place the Makefile driver knows about.
> - Have the Makefile driver call ocamltest with all result files for summarization: `ocamltest --summarize -args0 list-of-result-files`.
>
> This design keeps the current driver logic, in particular we can use
> `parallel` in the same way as before, with no parsing of results.

To give one example and although it's probably not very realistic, there
is an XML format for JUnit tests. If we produced output in this
JUnitformat, then we wcould rely on theparsers that exist for thisformat
and, for instance, JEnkinswould be able to take advantage of test
results.

**Comment by xavierleroy** [comment]
> What would be wrong with [using make -j to run tests in parallel] ?

For one thing, I don't know how the outputs of the tests would be interleaved on stdout.  GNU parallel does a good job of rearranging outputs in the same order as for a sequential run...  You might be able to do that with make wizardry, but do we need more make wizardry?

> In my opinion, it's rather amatter of producing the output of ocamltest in a format for whichthere are already parsers. 

Sounds like a nice approach, if you can find a good format with good tools.  No XML if possible...

I'll add my own feature wish: CI runs of the whole test suite should display nothing but the final summary and the output of the failed tests.  This would make it easier to investigate the failed tests, and reduce the size of email notifications.

**Comment by dra27** [comment]
FWIW, I implemented `make parallel` in `make` before discovering that GNU parallel actually worked in Cygwin - but that involves generating lots of file fragments and reconstructing the log and you don't get any output at all until the entire run has finished (IIRC). I don't think I deleted it, but I think GNU parallel is the better approach, too.

**Comment by dra27** [comment]
Ah, there we go: https://github.com/dra27/ocaml/commit/4d649cf7b95eb0c81c8d0e2652e68e62b3807de5. It looks like I addressed interleaving but didn't care about ordering.

**Comment by shindere** [comment]
David Allsopp (2022/03/10 08:21 -0800):
> Ah, there we go:
> https://github.com/dra27/ocaml/commit/4d649cf7b95eb0c81c8d0e2652e68e62b3807de5. It looks like I addressed interleaving but didn't care about ordering.

Sorry, I didn't pay attention and just submitted PR #11104.

**Comment by shindere** [comment]
This has been fixed by #11088 and #11100. Closing.
